KAFKA-3740: Enable configuration of RocksDBStores

dguy opened the pull request:
Add new config StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG to enable advanced
RocksDB users to override default RocksDB configuration

dguy committed:
Add new config StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG to enable advanced
RocksDB users to override default RocksDB configuration

dguy committed:
Add new config StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG to enable advanced
RocksDB users to override default RocksDB configuration

dguy commented:
@guozhangwang @norwood @theduderog - as discussed. Please take a look. Note: we don't currently limit memory usage, so that will be added once KIP-63 is done.

norwood commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -158,6 +162,12 @@ public RocksDBCacheEntry(V value, boolean isDirty) {
+        final Class<RocksDBConfigSetter> configSetterClass = (Class<RocksDBConfigSetter>) configs.get(StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG);
+        if (configSetterClass != null) {
+            final RocksDBConfigSetter configSetter = Utils.newInstance(configSetterClass);
+            configSetter.setConfig(name, options);
```
if you think theres something useful in `configs` for a `configSetter` to use, we could pass the `configs` in here as well so a `configSetter` wouldn't have to go get that data from elsewhere.

dguy replied:
Sure - makes sense. I'll add it. Thanks

theduderog replied:
Or pass the context object?

dguy replied:
I think the `configs` in this case as the `ProcessorContext` interface is quite broad, i.e, we don't want people calling `commit`, as an example

theduderog replied:
Cool.  Thanks.

dguy committed:
HOTFIX: fix compilation error due to merge of KAFKA-3812

Merge of KAFKA-3812 caused a compilation error in StreamThreadStateStoreProviderTest

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1641 from dguy/fix-compile-error

dguy committed:
add configs to RocksDbSetterConfig.setConfig

dguy committed:
Merge branch 'kafka-3740-listener' of github.com:dguy/kafka into kafka-3740-listener

norwood commented:
maybe not in this pr, but i think it would be good to eventually get this abstracted for any store type. e.g. if you have a mixture of rocks/inmemory stores you may want to be able to change the inmemory maxentries in a similar way.

theduderog commented on streams/src/main/java/org/apache/kafka/streams/state/RocksDBConfigSetter.java:
```diff
@@ -0,0 +1,41 @@
+     *  reduced due to memory constraints.
+     * </p>
+     * @param storeName     the name of the store being configured
+     * @param options       the Rocks DB optionse
```
typo `optionse`

theduderog commented on streams/src/main/java/org/apache/kafka/streams/state/RocksDBConfigSetter.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Set the rocks db options for the provided storeName.
+     * <p>
+     *  Note: Some options may be taken as a hint, i.e, {@link BlockBasedTableConfig#setBlockCacheSize(long)} might be
```
Since this isn't happening yet, should this comment be here?  It may not need to happen in the future either.  When memory constraints are in place, they can be included in the callback so that the user can make sure the constraints are not violated and therefore the settings are no longer hints.

dguy replied:
Sure - makes sense

dguy commented:
@norwood i think eventually you'll be able to supply your own custom built StoreSupplier via the DSL. So you'd be able to build to provide whatever settings you like to the InMemory or other store. The issue with Abstracting it is that what do you abstract it to? Config of an InMemory store is vastly different to an RocksDB store which will be vastly different to some Custom stores.

dguy committed:
fix comments

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java:
```diff
@@ -76,8 +77,9 @@ public void before() throws IOException {
         final String applicationId = "applicationId";
         properties.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
         properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        final String stateConfigDir = TestUtils.tempDirectory().getPath();
```
Should we delete this directory in `@after`?

dguy replied:
@guozhangwang Strictly speaking it doesn't really matter as the directory is deleteOnExit and there is a shutdown hook. However it is good practice to cleanup, so i've done it. Thanks

guozhangwang commented:
Minor comment about unit test, otherwise LGTM.

dguy committed:
cleanup tempDir in after

enothereska commented:
LGTM
